### PR TITLE
Add Domains V3 resource to CloudFoundryClient

### PR DESCRIPTION
**Pull request by LittleBaiBai**
[#1008]

### CODE REVIEW

**Comment by LittleBaiBai** [comment]
Hi @twoseat, do you have any feedbacks or concern on this PR?

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 import org.cloudfoundry.client.v2.applicationusageevents.ApplicationUsageEvents;
 import org.cloudfoundry.client.v2.blobstores.Blobstores;
 import org.cloudfoundry.client.v2.buildpacks.Buildpacks;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java]: This is a breaking change - ideally we would have thought ahead and called it `DomainsV2` from the beginning, but we are where we are, so this needs to remain unchanged. Introducing `DomainsV3` and `ReactorDomainsV3` is fine, even if it looks a little inconsistent.

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java [hunk 2]
```diff
@@ -67,7 +68,7 @@
 import org.cloudfoundry.reactor.client.v2.applicationusageevents.ReactorApplicationUsageEvents;
 import org.cloudfoundry.reactor.client.v2.blobstores.ReactorBlobstores;
 import org.cloudfoundry.reactor.client.v2.buildpacks.ReactorBuildpacks;
-import org.cloudfoundry.reactor.client.v2.domains.ReactorDomains;
+import org.cloudfoundry.reactor.client.v2.domains.ReactorDomainsV2;
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java [hunk 3]
```diff
@@ -169,8 +171,14 @@ public DeploymentsV3 deploymentsV3() {
...
-    public Domains domains() {
-        return new ReactorDomains(getConnectionContext(), getRootV2(), getTokenProvider(), getRequestTags());
+    public DomainsV2 domainsV2() {
+        return new ReactorDomainsV2(getConnectionContext(), getRootV2(), getTokenProvider(), getRequestTags());
+    }
+
+    @Override
+    @Value.Derived
+    public DomainsV3 domainsV3() {
+        return new ReactorDomainsV3(getConnectionContext(), getRootV3(), getTokenProvider(), getRequestTags());
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/_ReactorCloudFoundryClient.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java [hunk 4]
```diff
@@ -20,7 +20,7 @@
 import org.cloudfoundry.client.v2.domains.CreateDomainResponse;
 import org.cloudfoundry.client.v2.domains.DeleteDomainRequest;
 import org.cloudfoundry.client.v2.domains.DeleteDomainResponse;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java [hunk 5]
```diff
@@ -35,9 +35,9 @@
 import java.util.Map;
 
 /**
- * The Reactor-based implementation of {@link Domains}
+ * The Reactor-based implementation of {@link DomainsV2}
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java [hunk 6]
```diff
@@ -35,9 +35,9 @@
...
  */
-public final class ReactorDomains extends AbstractClientV2Operations implements Domains {
+public final class ReactorDomainsV2 extends AbstractClientV2Operations implements DomainsV2 {
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java [hunk 7]
```diff
@@ -47,7 +47,7 @@
      * @param tokenProvider     the {@link TokenProvider} to use when communicating with the server
      * @param requestTags       map with custom http headers which will be added to web request
      */
-    public ReactorDomains(ConnectionContext connectionContext, Mono<String> root, TokenProvider tokenProvider, Map<String, String> requestTags) {
+    public ReactorDomainsV2(ConnectionContext connectionContext, Mono<String> root, TokenProvider tokenProvider, Map<String, String> requestTags) {
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/domains/ReactorDomainsV3.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Override
+    public Mono<ListDomainsResponse> list(ListDomainsRequest request) {
+        return get(request, ListDomainsResponse.class, builder -> builder.pathSegment("domains"))
+            .checkpoint();
+    }
+
+    @Override
+    public Mono<ShareDomainResponse> share(ShareDomainRequest request) {
+//        return null;
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/domains/ReactorDomainsV3.java]: Remove comment

File: cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/domains/ReactorDomainsV3.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        return post(request, ShareDomainResponse.class, builder -> builder.pathSegment("domains", request.getDomainId(), "relationships", "shared_organizations"))
+            .checkpoint();
+    }
+
+    @Override
+    public Mono<Void> unshare(UnshareDomainRequest request) {
+//        return null;
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/v3/domains/ReactorDomainsV3.java]: Remove comment

File: cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/ReactorCloudFoundryClientTest.java [hunk 10]
```diff
@@ -56,8 +56,13 @@ public void builds() {
     }
 
     @Test
-    public void domains() {
-        assertThat(this.client.domains()).isNotNull();
+    public void domainsV2() {
+        assertThat(this.client.domainsV2()).isNotNull();
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/ReactorCloudFoundryClientTest.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2Test.java [hunk 11]
```diff
@@ -49,9 +49,9 @@
 import static io.netty.handler.codec.http.HttpResponseStatus.NO_CONTENT;
 import static io.netty.handler.codec.http.HttpResponseStatus.OK;
 
-public final class ReactorDomainsTest extends AbstractClientApiTest {
+public final class ReactorDomainsV2Test extends AbstractClientApiTest {
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2Test.java]: Breaking change (see above)

File: cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2Test.java [hunk 12]
```diff
@@ -49,9 +49,9 @@
...
 
-    private final ReactorDomains domains = new ReactorDomains(CONNECTION_CONTEXT, this.root, TOKEN_PROVIDER, Collections.emptyMap());
+    private final ReactorDomainsV2 domains = new ReactorDomainsV2(CONNECTION_CONTEXT, this.root, TOKEN_PROVIDER, Collections.emptyMap());
```
Comment (twoseat) [inline cloudfoundry-client-reactor/src/test/java/org/cloudfoundry/reactor/client/v2/domains/ReactorDomainsV2Test.java]: Breaking change (see above)

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/CloudFoundryClient.java [hunk 13]
```diff
@@ -20,7 +20,7 @@
 import org.cloudfoundry.client.v2.applicationusageevents.ApplicationUsageEvents;
 import org.cloudfoundry.client.v2.blobstores.Blobstores;
 import org.cloudfoundry.client.v2.buildpacks.Buildpacks;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/CloudFoundryClient.java]: Breaking change (see above)

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/CloudFoundryClient.java [hunk 14]
```diff
@@ -107,9 +108,14 @@
     DeploymentsV3 deploymentsV3();
 
     /**
-     * Main entry point to the Cloud Foundry Domains Client API
+     * Main entry point to the Cloud Foundry Domains V2 Client API
      */
-    Domains domains();
+    DomainsV2 domainsV2();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/CloudFoundryClient.java]: Breaking change (see above)

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v2/domains/DomainsV2.java [hunk 15]
```diff
@@ -19,9 +19,9 @@
 import reactor.core.publisher.Mono;
 
 /**
- * Main entry point to the Cloud Foundry Domains Client API
+ * Main entry point to the Cloud Foundry Domains V2 Client API
  */
-public interface Domains {
+public interface DomainsV2 {
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v2/domains/DomainsV2.java]: Breaking change (see above)

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/Domain.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    @JsonProperty("relationships")
+    @Nullable
+    public abstract DomainRelationships getRelationships();
+
+    /**
+     * The metadata
+     */
+    @JsonProperty("metadata")
+    @Nullable
+    public abstract Metadata getMetadata();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/Domain.java]: Move to alphabetical order

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/DomainsV3.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    Mono<CreateDomainResponse> create(CreateDomainRequest request);
+
+    /**
+     * Makes the <a href="http://v3-apidocs.cloudfoundry.org/version/3.77.0/index.html#update-a-domain">Create a domain</a> request
+     *
+     * @param request the Update a Domain request
+     * @return the response from the Update a Domain request
+     */
+    Mono<UpdateDomainResponse> update(UpdateDomainRequest request);
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/DomainsV3.java]: Alphabetical order

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_CreateDomainRequest.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    @JsonProperty("relationships")
+    @Nullable
+    abstract DomainRelationships getRelationships();
+
+    /**
+     * The metadata
+     */
+    @JsonProperty("metadata")
+    @Nullable
+    abstract Metadata getMetadata();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_CreateDomainRequest.java]: Alphabetical order

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ListDomainsRequest.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    @FilterParameter("organization_guids")
+    @Nullable
+    abstract List<String> getOwningOrganizationIds();
+
+    /**
+     * The metadata query
+     */
+    @FilterParameter("label_selector")
+    @Nullable
+    abstract String getLabelSelector();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ListDomainsRequest.java]: Alphabetical order

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainRequest.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+     * The domain id
+     */
+    @JsonIgnore
+    abstract String getDomainId();
+
+    /**
+     * The organizations the domain is shared to
+     */
+    @JsonProperty("data")
+    abstract List<Relationship> getData();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainRequest.java]: Alphabetical order

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainResponse.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * The response payload for the Share Service Instance operation
+ */
+@JsonDeserialize
+@Value.Immutable
+abstract class _ShareDomainResponse {
+
+    /**
+     * The space
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainResponse.java]: Seems this should be `The data`

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainResponse.java [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+import org.cloudfoundry.Nullable;
+import org.cloudfoundry.client.v3.Link;
+import org.cloudfoundry.client.v3.Relationship;
+import org.immutables.value.Value;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The response payload for the Share Service Instance operation
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/domains/_ShareDomainResponse.java]: I've made this copy/paste error so many times!

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/OrganizationsV3.java [hunk 23]
```diff
@@ -63,4 +63,12 @@
      */
     Mono<ListOrganizationsResponse> list(ListOrganizationsRequest request);
 
+    /**
+     * Makes <a href="http://v3-apidocs.cloudfoundry.org/version/3.77.0/index.html#list-domains-for-an-organization">List Organization Domains</a> request
+     *
+     * @param request the List Organization Domains request
+     * @return the response from the List Organization Domains request
+     */
+    Mono<ListOrganizationDomainsResponse> listDomains(ListOrganizationDomainsRequest request); //TODO needs tests
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/OrganizationsV3.java]: Does it still need tests?

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/_ListOrganizationDomainsRequest.java [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+    @FilterParameter("organization_guids")
+    @Nullable
+    abstract List<String> getOwningOrganizationIds();
+
+    /**
+     * The metadata query
+     */
+    @FilterParameter("label_selector")
+    @Nullable
+    abstract String getLabelSelector();
```
Comment (twoseat) [inline cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/_ListOrganizationDomainsRequest.java]: Alphabetical order

File: cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java [hunk 25]
```diff
@@ -19,7 +19,7 @@
 import org.cloudfoundry.client.CloudFoundryClient;
 import org.cloudfoundry.client.v2.applications.ApplicationsV2;
 import org.cloudfoundry.client.v2.buildpacks.Buildpacks;
-import org.cloudfoundry.client.v2.domains.Domains;
+import org.cloudfoundry.client.v2.domains.DomainsV2;
```
Comment (twoseat) [inline cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java]: Breaking change (see above)

File: cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java [hunk 26]
```diff
@@ -86,7 +86,7 @@
 
     protected final CloudFoundryClient cloudFoundryClient = mock(CloudFoundryClient.class, RETURNS_SMART_NULLS);
 
-    protected final Domains domains = mock(Domains.class, RETURNS_SMART_NULLS);
+    protected final DomainsV2 domainsV2 = mock(DomainsV2.class, RETURNS_SMART_NULLS);
```
Comment (twoseat) [inline cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java]: Breaking change (see above)

File: cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java [hunk 27]
```diff
@@ -149,7 +149,7 @@ public final void mockClient() {
         when(this.cloudFoundryClient.applicationsV2()).thenReturn(this.applications);
         when(this.cloudFoundryClient.applicationsV3()).thenReturn(this.applicationsV3);
         when(this.cloudFoundryClient.buildpacks()).thenReturn(this.buildpacks);
-        when(this.cloudFoundryClient.domains()).thenReturn(this.domains);
+        when(this.cloudFoundryClient.domainsV2()).thenReturn(this.domainsV2);
```
Comment (twoseat) [inline cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/AbstractOperationsTest.java]: Breaking change (see above)

File: cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/OrganizationsV3.java (hunk 23 above)
Comment (LittleBaiBai) [reply cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/organizations/OrganizationsV3.java:72]: Good catch! I added test but forgot to remove the TODO 😅

**Comment by LittleBaiBai** [comment]
@twoseat Thanks for your thorough review! I believe I have addressed all the comments you left and made sure no `DomainsV2` showed up in a global search. 

I also added integration tests, but I may need some help with getting one of them to pass - `DomainsTest.delete`. This test fails because `delete` is an asynchronous operation and the API spec declared the response to have job location url, but the CF instance I'm using is not returning it. So I wasn't able to wait until job is done as in some other integration test, then the assertion right after deletion fails. We could add retries or arbitrary wait time, but I'd love to hear your thoughts on this first.

**Comment by twoseat** [comment]
Thanks for the updates @LittleBaiBai - I'm working on it now, and have found what I guess is your domains issue (there's an undocumented `warnings` field returned as part of the `job`, and we're configured to choke on unexpected fields for test purposes). I've added something temporarily that makes the test work, but this issue (and possibly other V3 stuff) might have to wait until I get some feedback on what the field is supposed to contain.

**Pull request closed without merging**